File: typescript/src/criteria/greedy.ts
import type {FactorsType, PairByKeyType, PairType, ScalarType} from '../types';
import { combinations, unique } from '../lib';
import { Controller } from '../controller';

const getNumRemovablePairs = (indexes: Set<number>, incomplete: PairByKeyType, strengths: number[], exclude?: Set<ScalarType>) => {
  let num = 0;
  for (const s of strengths) {
    if (s === 2) {
      // Fast path for pairwise: avoid combinations/unique overhead.
      const arr = [...indexes];
      const len = arr.length;
      for (let i = 0; i < len - 1; i++) {
        const ai = arr[i];
        for (let j = i + 1; j < len; j++) {
          const key = ai * arr[j];
          if (incomplete.has(key) && (!exclude || !exclude.has(key))) {
            num++;
          }
        }
      }
    } else {
      for (let pair of combinations([...indexes], s)) {
        const key = unique(pair);
        if (incomplete.has(key) && (!exclude || !exclude.has(key))) {
          num++;
        }
      }
    }
  }
  return num;
};

export default function*<T extends FactorsType> (ctrl: Controller<T>): Generator<PairType> {
  const hasConstraints = (ctrl.options?.constraints?.length ?? 0) > 0;
  while (true) {
    let maxNumPairs: number | null = null;
    let efficientPair: PairType | null = null;

    // Compute the set of pairs already covered by the current row so
    // getNumRemovablePairs can exclude them. These pairs are still in
    // `incomplete` (consumption happens at yield time) but should not be
    // counted as "removable" — they are already guaranteed to be covered
    // by this row if it succeeds.
    const rowCovered = new Set<ScalarType>();
    if (ctrl.row.size > 0) {
      for (const s of ctrl.allStrengths) {
        for (const p of combinations([...ctrl.row.values()], s)) {
          rowCovered.add(unique(p));
        }
      }
    }

    for (const [pairKey, pair] of ctrl.incomplete.entries()) {
      const rowSize = ctrl.row.size;
      if (ctrl.isFilled(ctrl.row)) {
        break;
      }

      // Skip pairs already covered by the current row.
      if (rowCovered.has(pairKey)) continue;

      // Skip pairs that have been tried and failed for this row.
      if (ctrl.row.invalidPairs.has(pairKey)) continue;

      // Fast path: check compatibility without allocations
      const compat = rowSize === 0 ? pair.length : ctrl.isCompatible(pair);
      if (compat === null) {
        continue;
      }
      if (compat === 0) {
        continue;
      }

      // If constraints exist, do full storable check
      let storable = compat;
      if (hasConstraints) {
        const fullStorable = ctrl.storable(ctrl.getCandidate(pair));
        if (fullStorable === null) {
          continue;
        }
        if (fullStorable === 0) {
          continue;
        }
        storable = fullStorable;
      }

      const storableAbs = Math.abs(storable);
      const { tolerance = 0 } = ctrl.options!;

      const numPairs = getNumRemovablePairs(
        new Set([...ctrl.row.values(), ...pair]),
        ctrl.incomplete,
        ctrl.allStrengths,
        rowCovered,
      );

      if (numPairs + tolerance > rowSize * storableAbs) {
        efficientPair = pair;
        break;
      }
      if (maxNumPairs === null || maxNumPairs < numPairs) {
        maxNumPairs = numPairs;
        efficientPair = pair;
      }

    }
    if (efficientPair === null) {
      break;
    }
    yield efficientPair;
  }
};

File: python/covertable/criteria/greedy.py
from itertools import combinations


def get_num_removable_pairs(indexes, incomplete, strengths, exclude=None):
    num = 0
    arr = sorted(indexes)
    for s in strengths:
        if s == 2:
            # Fast path for pairwise: avoid combinations overhead
            length = len(arr)
            for i in range(length - 1):
                ai = arr[i]
                for j in range(i + 1, length):
                    key = (ai, arr[j])
                    if key in incomplete and (exclude is None or key not in exclude):
                        num += 1
        else:
            for pair in combinations(arr, s):
                if pair in incomplete and (exclude is None or pair not in exclude):
                    num += 1
    return num


def extract(ctrl):
    has_constraints = bool(ctrl._constraints)
    while True:
        max_num_pairs = None
        efficient_pair = None

        # Compute pairs already covered by current row
        row_covered = set()
        if ctrl.row:
            for s in ctrl.all_strengths:
                for p in combinations(sorted(ctrl.row.values()), s):
                    row_covered.add(tuple(p))

        for pair_key, pair in list(ctrl.incomplete.items()):
            row_size = len(ctrl.row)

            if ctrl.is_filled():
                break

            # Skip pairs already covered by current row
            if pair_key in row_covered:
                continue

            # Skip pairs that failed for this row
            if pair_key in ctrl.row.invalid_pairs:
                continue

            # Fast compatibility check
            compat = len(pair) if row_size == 0 else ctrl.is_compatible(pair)
            if compat is None:
                continue
            if compat == 0:
                continue

            # Full storable check if constraints exist
            storable = compat
            if has_constraints:
                full_storable = ctrl.storable(ctrl.get_candidate(pair))
                if full_storable is None:
                    continue
                if full_storable == 0:
                    continue
                storable = full_storable

            storable_abs = abs(storable)

            num_pairs = get_num_removable_pairs(
                set(list(ctrl.row.values()) + list(pair)),
                ctrl.incomplete,
                ctrl.all_strengths,
                row_covered,
            )

            if num_pairs + ctrl.tolerance > row_size * storable_abs:
                efficient_pair = pair
                break
            if max_num_pairs is None or max_num_pairs < num_pairs:
                max_num_pairs = num_pairs
                efficient_pair = pair

        if efficient_pair is None:
            break
        yield efficient_pair
